doc/dev/tmem.md

```diff
@@ -0,0 +1,1167 @@
+  auto out = ke.run({t0});
+  EXPECT_TRUE(at::equal(out[0], t0));
+} /*
+```
+
+In the above example, each CTA only has one warp, and this warp is split into
+TIDz, TIDy, and TIDx. The above kernel uses a loop of 2, where each iteration
+accesses a 32x1 box of the tensor memory. This is a valid 32x32b pattern
+.<!-- */ //-->\
+```cpp
+TEST_F(TMemTutorialR, WarpGroupXYZ) {
+  NOT_IMPLEMENTED
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigConcreteTensor({2, 8, 8, 2});
+  fusion.addInput(tv0);
+  auto tv1 = set(tv0);
+  auto tv2 = set(tv1);
+  auto tv3 = set(tv2);
+  auto tv4 = set(tv3);
+  fusion.addOutput(tv4);
+  tv2->setMemoryType(MemoryType::Tensor);
+
+  tv4->axis(0)->parallelize(ParallelType::TIDz);
+  tv4->axis(1)->parallelize(ParallelType::TIDy);
+  tv4->axis(2)->parallelize(ParallelType::TIDx);
+  scheduler_utils::parallelizeAllLike(tv4);
+
+  tv2->setTMemDimSepPos(3);
+
+  if constexpr (verbose) {
+    fusion.printKernel();
+  }
+
+  KernelExecutor ke;
+  ke.compile(&fusion);
+
+  at::Tensor t0 = at::rand({2, 8, 8, 2}, at::kCUDA);
+  auto out = ke.run({t0});
+  EXPECT_TRUE(at::equal(out[0], t0));
+} /*
+```
+
+In the above example, each CTA has one warp group. This entire warp group is
+accessing a whole column, with each warp accessing its subpartition of
+32 lanes. This is a valid 32x32b pattern.<!-- */ //-->\
+```cpp
+TEST_F(TMemTutorialR, WarpGroupXYColZ) {
+  NOT_IMPLEMENTED
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigConcreteTensor({8, 16, 8});
+  fusion.addInput(tv0);
+  auto tv1 = set(tv0);
+  auto tv2 = set(tv1);
+  auto tv3 = set(tv2);
+  auto tv4 = set(tv3);
+  fusion.addOutput(tv4);
+  tv2->setMemoryType(MemoryType::Tensor);
+
+  tv4->axis(0)->parallelize(ParallelType::TIDy);
+  tv4->axis(1)->parallelize(ParallelType::TIDx);
+  tv4->axis(2)->parallelize(ParallelType::TIDz);
+  scheduler_utils::parallelizeAllLike(tv4);
+
+  tv2->setTMemDimSepPos(2);
+
+  if constexpr (verbose) {
+    fusion.printKernel();
+  }
+
+  KernelExecutor ke;
+  ke.compile(&fusion);
+
+  at::Tensor t0 = at::rand({8, 16, 8}, at::kCUDA);
+  auto out = ke.run({t0});
+  EXPECT_TRUE(at::equal(out[0], t0));
+} /*
+```
+
+In the above example, each CTA has 8 warp groups, each warp group accesses a
+whole column. Warp group `i` is accessing column `i`.
+This is a valid 32x32b pattern.<!-- */ //-->\
+```cpp
+TEST_F(TMemTutorialR, WarpGroupXColYZ) {
+  NOT_IMPLEMENTED
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigConcreteTensor({128, 2, 2});
+  fusion.addInput(tv0);
+  auto tv1 = set(tv0);
+  auto tv2 = set(tv1);
+  auto tv3 = set(tv2);
+  auto tv4 = set(tv3);
+  fusion.addOutput(tv4);
+  tv2->setMemoryType(MemoryType::Tensor);
+
+  tv4->axis(0)->parallelize(ParallelType::TIDx);
+  tv4->axis(1)->parallelize(ParallelType::TIDy);
+  tv4->axis(2)->parallelize(ParallelType::TIDz);
+  scheduler_utils::parallelizeAllLike(tv4);
+
+  tv2->setTMemDimSepPos(2);
```

Comment:
Isn't this be 1?

Reply:
Oh sorry. You are right, it should be 1.